File: problem002.py
""" problem002.py:
By considering the terms in the Fibonacci sequence whose values do not exceed four million, find the sum of the even-valued terms. """
from functools import lru_cache


# This lru_cache does all the memoization work.
@lru_cache()
def fib(n):
    if n == 0:
        return 1
    elif n == 1:
        return 1
    else:
        return fib(n-1) + fib(n-2)


def naive(max_value):
    i = 0
    out = 0
    while True:
        fib_i = fib(i)
        if fib_i > max_value:  # Exceeds 4 million. Or whatever.
            break
        if fib_i % 2 == 0:  # Add even valued term.
            out += fib_i
        i += 1
    return out

if __name__ == "__main__":
    print(naive(4000000))

File: javascript/problem002.js
// problem002.js
// Fibonacci time!!

function fibonacci(num) {
    if (fibonacci.cache[num] == null) {
        fibonacci.cache[num] = fibonacci.cache[num-2] + fibonacci.cache[num-1];
    }
    return fibonacci.cache[num];
}

fibonacci.cache = {0: 1, 1: 1};

function main(max_value) {
    var i = 0;
    var out = 0;
    while (true) {
        var fib_i = fibonacci(i);
        if (fib_i > max_value) {
            break;
        } else if (fib_i %2 === 0) {
            out += fib_i;
        }
        i += 1;
    }
    return out;
}

console.log(main(4000000));

